File: USACO/2.3/money.cpp
#include <bits/stdc++.h>
using namespace std;
#define loop(i, a) for (int i = 0; i < a; i++)

long long dp[10001][25];
int V, N;
int coins[25];

/*
ID: 11074501
TASK: money
LANG: C++
*/


int main(){
	ofstream fout ("money.out");
	ifstream fin ("money.in");

	fin >> V >> N;

	memset(dp, 0, sizeof dp);

	int counter = 0;
	while(!fin.eof()){
		string line;
		getline(fin, line);

		istringstream iss(line);
		int num = 0;
		while (iss >> num){
			coins[counter] = num;
			counter++;
		}
	}

	for (int i = 0; i <= N; i += coins[0]){
		dp[i][0] = 1;
	}

	for (int i = 1; i < V; i++){
		for (int j = 0; j <= N; j++){
			dp[j][i] = dp[j][i-1];
		}


		for (int j = 0; j <= N; j++){
			int itself = 1;

			while( (j + (itself*coins[i])) <= N){
				if (dp[j][i-1] > 0){
					dp[j + (itself * coins[i])][i] += dp[j][i-1];
				}
				itself++;
			}
		}
	}
	
	fout << dp[N][V-1] << endl;
}

File: USACO/2.3/money.py
'''
ID: 11074501
TASK: money
LANG: PYTHON3
'''

with open('money.in', 'r') as f:
	thing = f.readline().split()
	V = int(thing[0])
	N = int(thing[1])

	coins = []
	for line in f:
		thing = line.split()
		for i in thing:
			coins.append(int(i))

	dp = [[0]*(N+1) for i in range(V)]

	for i in range(0,N+1,coins[0]):
		if i % coins[0] == 0:
			dp[0][i] = 1

	for i in range(1, V):
		for j in range(0, N+1):
			dp[i][j] = dp[i-1][j]

		for j in range(0, N+1):
			itself = 1

			while (j + (itself * coins[i])) <= N:
				if dp[i-1][j] > 0:
					dp[i][j + (itself * coins[i])] += dp[i-1][j]
				itself += 1
	'''
	for idx in range(V):
		print(coins[idx], [(j, dp[idx][j]) for j in range(len(dp[idx]))])
	print(dp[V-1][N])
	'''

	with open('money.out', 'w') as fo:
		fo.write(str(dp[V-1][N]) + '\n')